src/structures/timedcache.ts

```diff
@@ -0,0 +1,103 @@
+interface ITimedValue<V> {
+    value: V;
+    ts: number;
+}
+
+export class TimedCache<K, V> implements Map<K, V> {
+    private readonly  map: Map<K, ITimedValue<V>>;
+
+    public constructor(private readonly liveFor: number) {
+        this.map = new Map();
+    }
+
+    public clear(): void {
+        this.map.clear();
+    }
+
+    public delete(key: K): boolean {
+        return this.map.delete(key);
+    }
+
+    public forEach(callbackfn: (value: V, key: K, map: Map<K, V>) => void): void {
+        for (const item of this) {
+            callbackfn(item[1], item[0], this);
+        }
+    }
+
+    public get(key: K): V | undefined {
+        const v = this.map.get(key);
+        if (v) {
+            const val = this.filterV(v);
+            if (val !== undefined) {
+                return val;
+            }
+            this.map.delete(key);
+        }
+    }
+
+    public has(key: K): boolean {
+        return this.get(key) !== undefined;
+    }
+
+    public set(key: K, value: V): this {
+        this.map.set(key, {
+            ts: Date.now(),
+            value,
+        });
+        return this;
+    }
+
+    public get size(): number {
+        return this.map.size;
+    }
+
+    public [Symbol.iterator](): IterableIterator<[K, V]> {
+        let iterator: IterableIterator<[K, ITimedValue<V>]>;
+        return {
+            next: () => {
+                if (!iterator) {
+                    iterator = this.map.entries();
+                }
+                let item: IteratorResult<[K, ITimedValue<V>]>|undefined;
+                let filteredValue: V|undefined;
+                // Loop if => 1) no item/filteredvalue OR item && !filteredValue && not done
+                while (!item) {
+                    item = iterator.next();
+                    if (item.done) {
+                        break;
+                    }
+                    filteredValue = this.filterV(item.value[1]);
+                }
+                if (!item || filteredValue === undefined) {
```

Comment:
won't this prematurely stop iterating over the cache? Shouldn't we instead `while (!item && filteredValue === undefined)` ?

Reply:
Now this if-condition won't ever run, you could just remove it.

If it breaks due to `item.done` being true, then `if(!item)` wouldn't run, and with filteredValue it can't be undefined at the end of this